File: 5/Programmers/문자열압축.cpp
#include <string>
#include <vector>
using namespace std;

int solution(string s) {
  int answer = s.size();
  int count = 1;
  while (count <= s.size() / 2) {
    int idx = 0;
    string temp = "";
    while (idx <= s.size() - count) {
      int num = 1;
      string a = s.substr(idx, count);
      for (int i = idx + count; i <= s.size() - count; i += count) {
        string b = s.substr(i, count);
        if (a == b) {
          num++;
        } else {
          break;
        }
      }
      if (num != 1) {
        temp += to_string(num);
        temp += a;
        idx += (count * num);
      } else {
        temp += a;
        idx += count;
      }
    }
    if (idx != s.size()) {
      temp += s.substr(idx, s.size() - idx);
    }
    if (answer > temp.size()) {
      answer = temp.size();
    }
    count++;
  }
  return answer;
}

File: 2021/12/Programmers/문자열압축.py
def list_chunk(lst, n):
    return [lst[i : i + n] for i in range(0, len(lst), n)]


def solution(s):
    answer = len(s)

    num = len(s) // 2

    while num != 0:
        ans = ""
        temp = list_chunk(s, num)
        cnt = 1

        for idx, val in enumerate(temp):
            if idx == 0:
                continue

            if idx == len(temp) - 1:
                if val == temp[idx - 1]:
                    cnt = cnt + 1
                    ans = ans + str(cnt) + temp[idx - 1]
                else:
                    if cnt == 1:
                        ans = ans + temp[idx - 1]
                    else:
                        ans = ans + str(cnt) + temp[idx - 1]
                        cnt = 1
                    ans = ans + val
                break

            if val == temp[idx - 1]:
                cnt = cnt + 1
            else:
                if cnt == 1:
                    ans = ans + temp[idx - 1]
                else:
                    ans = ans + str(cnt) + temp[idx - 1]
                    cnt = 1

        if len(ans) < answer:
            answer = len(ans)
        num = num - 1
    return answer
